### PR TITLE
Fixes #2659 - Make console logs in browser configuration details easier to read

### LINKED ISSUE #2659: Make console logs in browser configuration details easier to read
Now that we have console logs coming in, we can think of ways to make them prettier.

screen shot 2018-10-16 at 4 27 47 pm
screen shot 2018-10-16 at 4 28 14 pm

cf. https://github.com/webcompat/webcompat.com/pull/2658

### PR DESCRIPTION
**Pull request by ksy36**
r? @karlcow @miketaylr

### CODE REVIEW

File: webcompat/templates/console-logs.html [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <title>Console logs | webcompat.com</title>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
```
Comment (miketaylr) [inline webcompat/templates/console-logs.html]: Maybe we should add a `<meta name="robots" content="noindex">` here?

File: tests/unit/test_console_logs.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+'''Tests for image upload API endpoint.'''
```
[the remark this answers is not in the document]
Comment (ksy36) [reply tests/unit/test_console_logs.py]: ah thanks for noticing. I've copied the file and didn't change that

File: webcompat/helpers.py [hunk 3]
```diff
@@ -412,13 +412,14 @@ def wrapped_func(*args, **kwargs):
                 file_path = FIXTURES_PATH + request.path + "." + checksum
             else:
                 file_path = FIXTURES_PATH + request.path
-            full_file_path = file_path + '.json'
-            if not os.path.exists(full_file_path):
-                print('Fixture expected at: {fix}'.format(fix=full_file_path))
+            if '.json' not in file_path:
```
[the remark this answers is not in the document]
Comment (ksy36) [reply webcompat/helpers.py]: thanks, I've updated that

File: webcompat/api/console_logs.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import os
+import uuid
+
+from flask import abort
+from flask import Blueprint
+from flask import request
+
+from webcompat import app
+
+console_logs = Blueprint('console_logs', __name__, url_prefix='/console_logs')
```
[the remark this answers is not in the document]
Comment (ksy36) [reply webcompat/api/console_logs.py]: Thanks for the feedback. I've moved console log saving to uploads blueprint, could you please take another look @miketaylr @karlcow

File: tests/unit/test_urls.py [hunk 5]
```diff
@@ -151,6 +151,11 @@ def test_contact(self):
         rv = self.app.get('/contact')
         self.assertEqual(rv.status_code, 200)
 
+    def test_console_logs(self):
+        """Test that /console_logs exists."""
+        rv = self.app.get('/console_logs/11/20/1c0a7174-2c15-480f-9cee-e6183e6a781b')  # noqa
```
Comment (karlcow) [inline tests/unit/test_urls.py]: note to self, we probably needs to create a console_logs directory

**Comment by miketaylr** [comment]
🎈 I'll deploy this to staging so we can play around with it.

**Comment by miketaylr** [comment]
@ksy36 I filed a test issue on staging:

https://staging.webcompat.com/issues/2274

Screen Shot 2019-12-10 at 12 57 03 PM

**Comment by miketaylr** [comment]
Screen Shot 2019-12-10 at 1 00 14 PM

Looks like we just need to sanitize the JSON, as it may contain HTML strings (and would be an XSS attack vector).

**Comment by karlcow** [comment]
@ksy36 

> Looks like we just need to sanitize the JSON, as it may contain HTML strings (and would be an XSS attack vector).

so that's what I wanted to check but @miketaylr beat me to it. Cool!
my hunch was motivated by 

https://github.com/webcompat/webcompat.com/blob/e12e205de140befc1c46b565314593cfed71005e/webcompat/helpers.py#L766-L774

Instead of True, False here the function should probably report a type, be Image or JSON (These could be strings). To return this specific type we need to validate it, the validation criteria needs to make it clear what we accept or not. 

Then there is the question that @miketaylr was pointing about sanitization. server side or client side?

**Comment by ksy36** [comment]
> Instead of True, False here the function should probably report a type, be Image or JSON (These could be strings). To return this specific type we need to validate it, the validation criteria needs to make it clear what we accept or not.
> 

For both of them validation happens later on, in their corresponding classes after an instance is created and before saving attempt. I wanted to separate these two paths (image/json) and have related methods contained within their classes. 

So `get_data_from_request` method just meant to indicate which class should be used based on request contents. I'll change True/False to return type as string (image/json) as @karlcow suggests, since now it's a bit confusing. 

And validation we have these:

For image (existing validation):
https://github.com/webcompat/webcompat.com/blob/e12e205de140befc1c46b565314593cfed71005e/webcompat/api/uploads.py#L68-L86

For json, it just checks that passed string is a valid JSON
https://github.com/webcompat/webcompat.com/blob/e12e205de140befc1c46b565314593cfed71005e/webcompat/api/uploads.py#L149-L153

Maybe we could use something like https://python-jsonschema.readthedocs.io/en/latest/ to validate json schema.

What do you think @karlcow ?

**Comment by ksy36** [comment]
For sanitization part, @miketaylr mentioned https://api.jquery.com/text/ will escape HTML tags. So I've tried it and it works ok, I'll test it a bit more :)

**Comment by karlcow** [comment]
> So `get_data_from_request` method just meant to indicate which class should be used based on request contents. I'll change True/False to return type as string (image/json) as @karlcow suggests, since now it's a bit confusing.

Good idea. 


> For image (existing validation):
> 
> https://github.com/webcompat/webcompat.com/blob/e12e205de140befc1c46b565314593cfed71005e/webcompat/api/uploads.py#L68-L86
> 
> For json, it just checks that passed string is a valid JSON
> 
> https://github.com/webcompat/webcompat.com/blob/e12e205de140befc1c46b565314593cfed71005e/webcompat/api/uploads.py#L149-L153


So for images we try to convert to a Pillow Object (and Pillow handles the validation by raising TypeError if it doesn't work)

Probably we should rename the `process_data` as `process_json`
The schema validation is an interesting idea, but I have the fear that we would put ourselves in an dependency maintenance nightmare. Aka each time devtools is changing the format we would have to update the schema. So probably not necessary right now, but open question.


Thanks a lot @ksy36

**Comment by ksy36** [comment]
So I'm thinking maybe a safer way for sanitization is to escape html tags before saving to the server. That way we don't have malicious code uploaded (in theory :)). 

I've tried this and it seems to be working good given that the structure of json we get from the reporter is consistent:

```python
    import html
    ....
    def process_json(self, data):
        try:
            d = json.loads(data)
            new_set = []
            for item in d:
                new_dict = {}
                for key, value in item.items():
                    if isinstance(value, list):
                        new_dict[key] = ', '.join(list(map(lambda x: html.escape(x), value)))
                    else:
                        new_dict[key] = html.escape(value)
                new_set.append(new_dict)
            return new_set
        except ValueError:
            abort(400)
```
So it converts strings like this 

`Loading failed for the <script> with source https://...ads.js.`

To be:
`Loading failed for the &lt;script&gt; with source https://...ads.js.`

This involves some extra processing on the python side though. Do you have any thoughts on this @karlcow @miketaylr ?

**Comment by miketaylr** [comment]
I agree escaping before we save to disk is a better idea - we don't want to accidentally regress the client side escape...

Looking at https://github.com/webcompat/webcompat.com/pull/3103/files#diff-fbcd0a4c752e94ffddefb108ac14160fR196

I found the following answer on stackoverflow: https://stackoverflow.com/a/27382959

Which points to `htmlsafe_dumps` which comes with the flask json module: https://github.com/pallets/flask/blob/1351d0a56580df36872b466eb245e7634c20dab5/src/flask/json/__init__.py#L264. 

Can we just use that instead of `dumps`?

**Comment by miketaylr** [comment]
Oops, wrong link!

https://github.com/webcompat/webcompat.com/pull/3103/files#diff-fbcd0a4c752e94ffddefb108ac14160fR166

and https://github.com/pallets/flask/blob/1351d0a56580df36872b466eb245e7634c20dab5/src/flask/json/__init__.py#L301 makes more sense.

**Comment by ksy36** [comment]
Moving json rendering to the server (with jinja) helped with sanitization. So a <script> element, for example, is returned as `&lt;script&gt;` with no extra processing on our side. Also the js script is no longer needed :)

**Pull request merged**